Test/src/MergeSort.java

```diff
@@ -0,0 +1,78 @@
+/* Time complexity - N log N
+ * Space Complexity - N Log N
+ */
+
+public class MergeSort {
+	public static void main(String args[]){
+		
+		int[] unsorted = new int[] {0,9,8,7,6,5,4,3,2,1,0,1,2};
+		int[] sortedArray = new int[unsorted.length];
+		
+		sortedArray = mergesort(unsorted);
+		for(int i = 0; i < sortedArray.length; i++)
+			System.out.println( sortedArray[i]);
+		
+	}
+	
+	public static int[] mergesort(int[] unsorted){
+		int lengthOfUnsortedArray = unsorted.length;
+		if(lengthOfUnsortedArray <= 1){
+			return unsorted;
+		}
+		int arrayMid = lengthOfUnsortedArray/2;
+		int lengthOfUnsortedLeft=arrayMid;
+		int[]  leftUnsortedArray = new int[lengthOfUnsortedLeft];
+		
+		int lengthOfUnsortedRight=(lengthOfUnsortedArray-arrayMid);
+		int[]  rightUnsortedArray = new int[lengthOfUnsortedRight];
+		int k = 0;
+		
+		leftUnsortedArray=copyArray(unsorted, k, lengthOfUnsortedLeft);
+		rightUnsortedArray=copyArray(unsorted, k + lengthOfUnsortedLeft, lengthOfUnsortedRight);
```

Comment:
is it possible to skip creating a copy of left un sorted and right unsorted ?

Can you just send array, startIndex, endIndex as arguments and avoid this copy ?